server/lib/chat/chat-service.test.ts

```diff
@@ -0,0 +1,1505 @@
+        client1,
+        getChannelUserPath(shieldBatteryChannel.id, user1.id),
+        undefined,
+      )
+    })
+  })
+
+  describe('joinChannel', () => {
+    const addUserToChannelMock = asMockedFunction(addUserToChannel)
+    const addMessageToChannelMock = asMockedFunction(addMessageToChannel)
+    const createChannelMock = asMockedFunction(createChannel)
+
+    beforeEach(async () => {
+      // NOTE(2Pac): We pre-join user2 to the channel so we can check it sees correct websocket
+      // messages from user1 joining.
+      await joinUserToChannel(
+        user2,
+        shieldBatteryChannel,
+        user2ShieldBatteryChannelEntry,
+        joinUser2ShieldBatteryChannelMessage,
+      )
+
+      asMockedFunction(findUserById).mockResolvedValue(user1)
+      asMockedFunction(findChannelByName).mockResolvedValue(shieldBatteryChannel)
+      asMockedFunction(getUserChannelEntryForUser).mockResolvedValue(null)
+      asMockedFunction(isUserBannedFromChannel).mockResolvedValue(false)
+      asMockedFunction(countBannedIdentifiersForChannel).mockResolvedValue(0)
+      addUserToChannelMock.mockResolvedValue(user1ShieldBatteryChannelEntry)
+      addMessageToChannelMock.mockResolvedValue(joinUser1ShieldBatteryChannelMessage)
+      createChannelMock.mockResolvedValue(shieldBatteryChannel)
+    })
+
+    test("should throw if user doesn't exist", async () => {
+      asMockedFunction(findUserById).mockResolvedValue(undefined)
+
+      await expect(
+        chatService.joinChannel(shieldBatteryChannel.name, user1.id),
+      ).rejects.toThrowErrorMatchingInlineSnapshot(`"User doesn't exist"`)
+    })
+
+    test('should throw if user is banned', async () => {
+      asMockedFunction(isUserBannedFromChannel).mockResolvedValue(true)
+
+      await expect(
+        chatService.joinChannel(shieldBatteryChannel.name, user1.id),
+      ).rejects.toThrowErrorMatchingInlineSnapshot(`"User is banned"`)
+    })
+
+    test('should throw if smurf account is banned', async () => {
+      asMockedFunction(countBannedIdentifiersForChannel).mockResolvedValue(
+        MIN_IDENTIFIER_MATCHES + 1,
+      )
+
+      await expect(
+        chatService.joinChannel(shieldBatteryChannel.name, user1.id),
+      ).rejects.toThrowErrorMatchingInlineSnapshot(`"User is banned"`)
+    })
+
+    test('works when channel already exists', async () => {
+      await chatService.joinChannel(shieldBatteryChannel.name, user1.id)
+
+      expect(addUserToChannelMock).toHaveBeenCalledWith(user1.id, shieldBatteryChannel.id, dbClient)
+      expect(addMessageToChannelMock).toHaveBeenCalledWith(
+        user1.id,
+        shieldBatteryChannel.id,
+        { type: ServerChatMessageType.JoinChannel },
+        dbClient,
+      )
+      expect(client2.publish).toHaveBeenCalledWith(getChannelPath(shieldBatteryChannel.id), {
+        action: 'join2',
+        user: user1,
+        message: {
+          id: joinUser1ShieldBatteryChannelMessage.msgId,
```

Comment:
You do this like a million times throughout this file, I'd recommend instead writing a simple method that returns this `message` structure given the DB input, it'd reduce a lot of the noise in these tests